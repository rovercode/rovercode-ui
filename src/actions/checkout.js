// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const CREATE_CHECKOUT_SESSION = 'CREATE_CHECKOUT_SESSION';
export const CREATE_CHECKOUT_SESSION_PENDING = `${CREATE_CHECKOUT_SESSION}_PENDING`;
export const CREATE_CHECKOUT_SESSION_FULFILLED = `${CREATE_CHECKOUT_SESSION}_FULFILLED`;
export const CREATE_CHECKOUT_SESSION_REJECTED = `${CREATE_CHECKOUT_SESSION}_REJECTED`;

export const createCheckoutSession = (userId, lineItems, successURL, cancelURL, collectShippingAddress, xhrOptions) => ({
  type: CREATE_CHECKOUT_SESSION,
  payload: axios.post(`${SUBSCRIPTION_SERVICE}/api/v1/checkout/start/`, { // eslint-disable-line no-undef
    'id': userId,
    'line_items': lineItems,
    'success_url': successURL,
    'cancel_url': cancelURL,
    'collect_shipping_address': collectShippingAddress,
  }, xhrOptions)
    .then(({ data }) => (
      data
    )),
});