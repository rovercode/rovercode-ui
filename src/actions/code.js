// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const FETCH_PROGRAM = 'FETCH_PROGRAM';
export const FETCH_PROGRAM_PENDING = `${FETCH_PROGRAM}_PENDING`;
export const FETCH_PROGRAM_FULFILLED = `${FETCH_PROGRAM}_FULFILLED`;
export const FETCH_PROGRAM_REJECTED = `${FETCH_PROGRAM}_REJECTED`;
export const SAVE_PROGRAM = 'SAVE_PROGRAM';
export const SAVE_PROGRAM_PENDING = `${SAVE_PROGRAM}_PENDING`;
export const SAVE_PROGRAM_FULFILLED = `${SAVE_PROGRAM}_FULFILLED`;
export const SAVE_PROGRAM_REJECTED = `${SAVE_PROGRAM}_REJECTED`;
export const CREATE_PROGRAM = 'CREATE_PROGRAM';
export const CREATE_PROGRAM_PENDING = `${CREATE_PROGRAM}_PENDING`;
export const CREATE_PROGRAM_FULFILLED = `${CREATE_PROGRAM}_FULFILLED`;
export const CREATE_PROGRAM_REJECTED = `${CREATE_PROGRAM}_REJECTED`;

export const UPDATE_JSCODE = 'UPDATE_JSCODE';
export const UPDATE_XMLCODE = 'UPDATE_XMLCODE';
export const CHANGE_EXECUTION_STATE = 'CHANGE_EXECUTION_STATE';
export const CHANGE_NAME = 'CHANGE_NAME';
export const CHANGE_NAME_PENDING = `${CHANGE_NAME}_PENDING`;
export const CHANGE_NAME_FULFILLED = `${CHANGE_NAME}_FULFILLED`;
export const CHANGE_NAME_REJECTED = `${CHANGE_NAME}_REJECTED`;
export const CHANGE_PROGRAM_TAGS = 'CHANGE_PROGRAM_TAGS';
export const CHANGE_PROGRAM_TAGS_PENDING = `${CHANGE_PROGRAM_TAGS}_PENDING`;
export const CHANGE_PROGRAM_TAGS_FULFILLED = `${CHANGE_PROGRAM_TAGS}_FULFILLED`;
export const CHANGE_PROGRAM_TAGS_REJECTED = `${CHANGE_PROGRAM_TAGS}_REJECTED`;
export const CHANGE_ID = 'CHANGE_ID';
export const CHANGE_READ_ONLY = 'CHANGE_READ_ONLY';
export const CLEAR_PROGRAM = 'CLEAR_PROGRAM';

// Execution States
export const EXECUTION_RUN = 1;
export const EXECUTION_STEP = 2;
export const EXECUTION_STOP = 3;
export const EXECUTION_RESET = 4;

// action creators
export const updateJsCode = (jsCode) => ({
  type: UPDATE_JSCODE,
  payload: jsCode,
});

export const updateXmlCode = (xmlCode) => ({
  type: UPDATE_XMLCODE,
  payload: xmlCode,
});

export const changeExecutionState = (state) => ({
  type: CHANGE_EXECUTION_STATE,
  payload: state,
});

export const changeName = (id, name, xhroptions) => ({
  type: CHANGE_NAME,
  payload: axios.patch(`/api/v1/block-diagrams/${id}/`, {
    name,
  }, xhroptions)
    .then(({ data }) => (
      data
    )),
});

export const changeId = (id) => ({
  type: CHANGE_ID,
  payload: id,
});

export const fetchProgram = (id, xhroptions) => ({
  type: FETCH_PROGRAM,
  payload: axios.get(`/api/v1/block-diagrams/${id}/`, xhroptions)
    .then(({ data }) => (
      data
    )),
});

export const saveProgram = (id, content, name, lesson, xhroptions) => ({
  type: SAVE_PROGRAM,
  payload: axios.put(`/api/v1/block-diagrams/${id}/`, {
    content,
    name,
    lesson,
  }, xhroptions)
    .then(({ data }) => (
      data
    )),
});

export const createProgram = (name, xhroptions) => ({
  type: CREATE_PROGRAM,
  payload: axios.post('/api/v1/block-diagrams/', {
    name,
    content: '<xml></xml>',
  }, xhroptions)
    .then(({ data }) => (
      data
    )),
});

export const changeProgramTags = (id, tags, xhroptions) => ({
  type: CHANGE_PROGRAM_TAGS,
  payload: axios.patch(`/api/v1/block-diagrams/${id}/`, {
    owner_tags: tags,
  }, xhroptions)
    .then(({ data }) => (
      data
    )),
});

export const changeReadOnly = (isReadOnly) => ({
  type: CHANGE_READ_ONLY,
  payload: isReadOnly,
});

export const clearProgram = () => ({
  type: CLEAR_PROGRAM,
});
