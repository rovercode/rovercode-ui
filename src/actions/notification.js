// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const SHOW_NOTIFICATION = 'SHOW_NOTIFICATION';
export const CLEAR_NOTIFICATION = 'CLEAR_NOTIFICATION';

// action creators
export const showNotification = (message, duration, severity) => ({
  type: SHOW_NOTIFICATION,
  payload: {
    message,
    duration,
    severity,
  },
});

export const clearNotification = () => ({
  type: CLEAR_NOTIFICATION,
});
