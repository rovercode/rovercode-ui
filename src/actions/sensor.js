// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const CHANGE_LEFT_SENSOR_STATE = 'CHANGE_LEFT_SENSOR_STATE';
export const CHANGE_RIGHT_SENSOR_STATE = 'CHANGE_RIGHT_SENSOR_STATE';
export const CHANGE_LIGHT_SENSOR_READINGS = 'CHANGE_LIGHT_SENSOR_READINGS';
export const CHANGE_LINE_SENSOR_READINGS = 'CHANGE_LINE_SENSOR_READINGS';
export const CHANGE_DISTANCE_SENSOR_READING = 'CHANGE_DISTANCE_SENSOR_READING';
export const CHANGE_BATTERY_VOLTAGE_READING = 'CHANGE_BATTERY_VOLTAGE_READING';

// Sensor States
export const COVERED = 1;
export const NOT_COVERED = 2;

// action creators
export const changeLeftSensorState = (state) => ({
  type: CHANGE_LEFT_SENSOR_STATE,
  payload: state,
});

export const changeRightSensorState = (state) => ({
  type: CHANGE_RIGHT_SENSOR_STATE,
  payload: state,
});

export const changeLightSensorReadings = (leftReading, rightReading) => ({
  type: CHANGE_LIGHT_SENSOR_READINGS,
  payload: { leftReading, rightReading },
});

export const changeLineSensorReadings = (leftReading, rightReading) => ({
  type: CHANGE_LINE_SENSOR_READINGS,
  payload: { leftReading, rightReading },
});

export const changeDistanceSensorReading = (distanceReading) => ({
  type: CHANGE_DISTANCE_SENSOR_READING,
  payload: distanceReading,
});

export const changeBatteryVoltageReading = (voltageReading) => ({
  type: CHANGE_BATTERY_VOLTAGE_READING,
  payload: voltageReading,
});
