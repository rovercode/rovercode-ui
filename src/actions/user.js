// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions
import axios from 'axios';
import jwtDecode from 'jwt-decode';

export const UPDATE_USER = 'UPDATE_USER';
export const EDIT_USER_USERNAME = 'EDIT_USER_USERNAME';
export const EDIT_USER_USERNAME_PENDING = `${EDIT_USER_USERNAME}_PENDING`;
export const EDIT_USER_USERNAME_FULFILLED = `${EDIT_USER_USERNAME}_FULFILLED`;
export const EDIT_USER_USERNAME_REJECTED = `${EDIT_USER_USERNAME}_REJECTED`;
export const EDIT_USER_PASSWORD = 'EDIT_USER_PASSWORD';
export const EDIT_USER_PASSWORD_PENDING = `${EDIT_USER_PASSWORD}_PENDING`;
export const EDIT_USER_PASSWORD_FULFILLED = `${EDIT_USER_PASSWORD}_FULFILLED`;
export const EDIT_USER_PASSWORD_REJECTED = `${EDIT_USER_PASSWORD}_REJECTED`;
export const EDIT_USER_SHOW_GUIDE = 'EDIT_USER_SHOW_GUIDE';
export const EDIT_USER_SHOW_GUIDE_PENDING = `${EDIT_USER_SHOW_GUIDE}_PENDING`;
export const EDIT_USER_SHOW_GUIDE_FULFILLED = `${EDIT_USER_SHOW_GUIDE}_FULFILLED`;
export const EDIT_USER_SHOW_GUIDE_REJECTED = `${EDIT_USER_SHOW_GUIDE}_REJECTED`;
export const FETCH_USER_STATS = 'FETCH_USER_STATS';
export const FETCH_USER_STATS_PENDING = `${FETCH_USER_STATS}_PENDING`;
export const FETCH_USER_STATS_FULFILLED = `${FETCH_USER_STATS}_FULFILLED`;
export const FETCH_USER_STATS_REJECTED = `${FETCH_USER_STATS}_REJECTED`;
export const REFRESH_SESSION = 'REFRESH_SESSION';
export const REFRESH_SESSION_PENDING = `${REFRESH_SESSION}_PENDING`;
export const REFRESH_SESSION_FULFILLED = `${REFRESH_SESSION}_FULFILLED`;
export const REFRESH_SESSION_REJECTED = `${REFRESH_SESSION}_REJECTED`;

// action creators
export const updateUser = (data) => ({
  type: UPDATE_USER,
  payload: data,
});

export const refreshSession = (cookies) => ({
  type: REFRESH_SESSION,
  payload: axios.post('/api/api-token-refresh/', {
    token: cookies.get('auth_jwt'),
  })
    .then(({ data }) => (
      jwtDecode(data.token)
    )),
});

export const editUserUsername = (username, xhrOptions) => ({
  type: EDIT_USER_USERNAME,
  payload: axios.put('/jwt/auth/user/', { username }, xhrOptions)
    .then(({ data }) => (
      data
    )),
});

export const editUserPassword = (password, xhrOptions) => ({
  type: EDIT_USER_PASSWORD,
  payload: axios.post('/jwt/auth/password/change/', {
    new_password1: password,
    new_password2: password,
  }, xhrOptions)
    .then(({ data }) => (
      data
    )),
});

export const editUserShowGuide = (id, show, xhrOptions) => ({
  type: EDIT_USER_SHOW_GUIDE,
  payload: axios.patch(`/api/v1/users/${id}/`, { show_guide: show }, xhrOptions)
    .then(({ data }) => (
      data
    )),
});

export const fetchUserStats = (id, xhroptions) => ({
  type: FETCH_USER_STATS,
  payload: axios.get(`/api/v1/users/${id}/stats/`, xhroptions)
    .then(({ data }) => (
      data
    )),
});
