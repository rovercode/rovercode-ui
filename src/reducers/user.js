import {
  UPDATE_USER,
  EDIT_USER_USERNAME_PENDING,
  EDIT_USER_USERNAME_FULFILLED,
  EDIT_USER_USERNAME_REJECTED,
  EDIT_USER_PASSWORD_PENDING,
  EDIT_USER_PASSWORD_FULFILLED,
  EDIT_USER_PASSWORD_REJECTED,
  EDIT_USER_SHOW_GUIDE_PENDING,
  EDIT_USER_SHOW_GUIDE_FULFILLED,
  EDIT_USER_SHOW_GUIDE_REJECTED,
  FETCH_USER_STATS_PENDING,
  FETCH_USER_STATS_FULFILLED,
  FETCH_USER_STATS_REJECTED,
  REFRESH_SESSION_PENDING,
  REFRESH_SESSION_FULFILLED,
  REFRESH_SESSION_REJECTED,
} from '../actions/user';

export default function user(
  state = {
    user_id: null,
    username: null,
    email: null,
    exp: null,
    showGuide: true,
    tier: 1,
    stats: null,
    isSocial: false,
    isEditingUsername: false,
    isEditingPassword: false,
    isEditingShowGuide: false,
    isFetchingStats: false,
    isRefreshingSession: false,
    editUsernameError: null,
    editPasswordError: null,
    editShowGuideError: null,
    fetchStatsError: null,
    refreshSessionError: null,
  },
  action,
) {
  switch (action.type) {
    case UPDATE_USER:
      return {
        ...state,
        user_id: action.payload.user_id,
        username: action.payload.username,
        email: action.payload.email,
        exp: action.payload.exp,
        showGuide: action.payload.show_guide,
        isSocial: action.payload.isSocial,
        tier: action.payload.tier,
      };
    case EDIT_USER_USERNAME_PENDING:
      return {
        ...state,
        isEditingUsername: true,
      };
    case EDIT_USER_USERNAME_FULFILLED:
      return {
        ...state,
        isEditingUsername: false,
        username: action.payload.username,
        email: action.payload.email,
      };
    case EDIT_USER_USERNAME_REJECTED:
      return {
        ...state,
        isEditingUsername: false,
        editUsernameError: action.payload,
      };
    case EDIT_USER_PASSWORD_PENDING:
      return {
        ...state,
        isEditingPassword: true,
      };
    case EDIT_USER_PASSWORD_FULFILLED:
      return {
        ...state,
        isEditingPassword: false,
      };
    case EDIT_USER_PASSWORD_REJECTED:
      return {
        ...state,
        isEditingPassword: false,
        editPasswordError: action.payload,
      };
    case EDIT_USER_SHOW_GUIDE_PENDING:
      return {
        ...state,
        isEditingShowGuide: true,
      };
    case EDIT_USER_SHOW_GUIDE_FULFILLED:
      return {
        ...state,
        isEditingShowGuide: false,
        showGuide: action.payload.show_guide,
      };
    case EDIT_USER_SHOW_GUIDE_REJECTED:
      return {
        ...state,
        isEditingShowGuide: false,
        editShowGuideError: action.payload,
      };
    case FETCH_USER_STATS_PENDING:
      return {
        ...state,
        isFetchingStats: true,
      };
    case FETCH_USER_STATS_FULFILLED:
      return {
        ...state,
        isFetchingStats: false,
        stats: action.payload,
      };
    case FETCH_USER_STATS_REJECTED:
      return {
        ...state,
        isFetchingStats: false,
        fetchStatsError: action.payload,
      };
    case REFRESH_SESSION_FULFILLED:
      return {
        ...state,
        isRefreshingSession: false,
        user_id: action.payload.user_id,
        username: action.payload.username,
        email: action.payload.email,
        exp: action.payload.exp,
        isSocial: action.payload.isSocial,
        tier: action.payload.tier,
      };
    case REFRESH_SESSION_PENDING:
      return {
        ...state,
        isRefreshingSession: true,
      };
    case REFRESH_SESSION_REJECTED:
      return {
        ...state,
        isRefreshingSession: false,
        refreshSessionError: action.payload,
      };
    default:
      return state;
  }
}
