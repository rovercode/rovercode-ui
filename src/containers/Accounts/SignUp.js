import React, { Component } from 'react';
import { Redirect } from 'react-router';
import { Link as RouterLink } from 'react-router-dom';
import { Lock, Mail, Person } from '@material-ui/icons';
import {
  Button,
  Grid,
  InputAdornment,
  Link,
  List,
  ListItem,
  ListItemText,
  TextField,
  Typography,
} from '@material-ui/core';
import { Alert } from '@material-ui/lab';
import { FormattedMessage, injectIntl } from 'react-intl';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withCookies, Cookies } from 'react-cookie';
import jwtDecode from 'jwt-decode';
import axios from 'axios';

import { updateUser as actionUpdateUser } from '@/actions/user';

const mapDispatchToProps = (dispatch) => ({
  updateUser: (data) => dispatch(actionUpdateUser({ ...data, isSocial: false })),
});

class SignUp extends Component {
  constructor(props) {
    super(props);

    this.state = {
      username: '',
      email: '',
      password1: '',
      password2: '',
      usernameError: null,
      emailError: null,
      password1Error: null,
      password2Error: null,
      nonFieldError: null,
      success: false,
    };
  }

  handleEmailChange = (event) => {
    this.setState({
      email: event.target.value,
    });
  }

  handlePassword1Change = (event) => {
    this.setState({
      password1: event.target.value,
    });
  }

  handlePassword2Change = (event) => {
    this.setState({
      password2: event.target.value,
    });
  }

  handleUsernameChange = (event) => {
    this.setState({
      username: event.target.value,
    });
  }

  errorMessage = () => {
    const {
      usernameError,
      emailError,
      password1Error,
      password2Error,
      nonFieldError,
    } = this.state;

    const errors = [
      usernameError,
      emailError,
      password1Error,
      password2Error,
      nonFieldError,
    ];

    const activeErrors = errors.filter(Boolean);

    if (activeErrors.length === 0) {
      return (null);
    }

    const messages = [];

    activeErrors.forEach((error) => {
      error.forEach((message) => {
        messages.push(message);
      });
    });

    /* eslint react/no-array-index-key: 0 */
    return (
      <Grid item>
        <Alert variant="outlined" severity="error">
          <List dense>
            {
              messages.map((message, i) => (
                <ListItem key={i}>
                  <ListItemText>
                    {`- ${message}`}
                  </ListItemText>
                </ListItem>
              ))
            }
          </List>
        </Alert>
      </Grid>
    );
  }

  signUp = (e) => {
    const { cookies, updateUser } = this.props;
    const {
      username,
      email,
      password1,
      password2,
    } = this.state;

    e.preventDefault();

    return axios.post('/jwt/auth/registration/', {
      username,
      email,
      password1,
      password2,
    })
      .then((response) => {
        updateUser(jwtDecode(response.data.access_token));
        cookies.set('auth_jwt', response.data.access_token, { path: '/' });
        this.setState({
          success: true,
        });
      })
      .catch((error) => {
        this.setState({
          usernameError: error.response.data.username,
          emailError: error.response.data.email,
          password1Error: error.response.data.password1,
          password2Error: error.response.data.password2,
          nonFieldError: error.response.data.non_field_errors,
        });
      });
  }

  render() {
    const { intl } = this.props;
    const {
      success,
      usernameError,
      emailError,
      password1Error,
      password2Error,
    } = this.state;

    const passwordPlaceholder = intl.formatMessage({
      id: 'app.signup.password',
      description: 'Placeholder for password entry',
      defaultMessage: 'Password',
    });

    const usernamePlaceholder = intl.formatMessage({
      id: 'app.signup.username',
      description: 'Placeholder for username entry',
      defaultMessage: 'Username',
    });

    const emailPlaceholder = intl.formatMessage({
      id: 'app.signup.email',
      description: 'Placeholder for email entry',
      defaultMessage: 'Email',
    });

    return (
      <>
        {
          success ? (
            <Redirect to="/" />
          ) : (null)
        }
        <Grid container direction="column" justify="center" alignItems="center" spacing={4}>
          <Grid item>
            <Typography variant="h4">
              <FormattedMessage
                id="app.signup.header"
                description="Header for sign up"
                defaultMessage="Sign Up"
              />
            </Typography>
          </Grid>
          <Grid item>
            <Typography>
              <FormattedMessage
                id="app.signup.check_1"
                description="First part of Checking if the user already has an account"
                defaultMessage="Already have an account? Then please"
              />
              {' '}
              <Link component={RouterLink} to="/accounts/login">
                <FormattedMessage
                  id="app.signup.check_2"
                  description="Second part of Checking if the user already has an account"
                  defaultMessage="sign in."
                />
              </Link>
            </Typography>
          </Grid>
          {this.errorMessage()}
          <Grid item>
            <form onSubmit={this.signUp}>
              <Grid item>
                <TextField
                  variant="outlined"
                  margin="dense"
                  fullWidth
                  required
                  InputProps={{
                    startAdornment: (
                      <InputAdornment position="start">
                        <Person />
                      </InputAdornment>
                    ),
                  }}
                  placeholder={usernamePlaceholder}
                  onChange={this.handleUsernameChange}
                  error={!!usernameError}
                />
              </Grid>
              <Grid item>
                <TextField
                  variant="outlined"
                  margin="dense"
                  fullWidth
                  required
                  InputProps={{
                    startAdornment: (
                      <InputAdornment position="start">
                        <Mail />
                      </InputAdornment>
                    ),
                  }}
                  placeholder={emailPlaceholder}
                  onChange={this.handleEmailChange}
                  error={!!emailError}
                />
              </Grid>
              <Grid item>
                <TextField
                  variant="outlined"
                  margin="dense"
                  fullWidth
                  required
                  type="password"
                  InputProps={{
                    startAdornment: (
                      <InputAdornment position="start">
                        <Lock />
                      </InputAdornment>
                    ),
                  }}
                  placeholder={passwordPlaceholder}
                  onChange={this.handlePassword1Change}
                  error={!!password1Error}
                />
              </Grid>
              <Grid item>
                <TextField
                  variant="outlined"
                  margin="dense"
                  fullWidth
                  required
                  type="password"
                  InputProps={{
                    startAdornment: (
                      <InputAdornment position="start">
                        <Lock />
                      </InputAdornment>
                    ),
                  }}
                  placeholder={passwordPlaceholder}
                  onChange={this.handlePassword2Change}
                  error={!!password2Error}
                />
              </Grid>
              <Grid item>
                <Button variant="contained" color="primary" type="submit">
                  <FormattedMessage
                    id="app.signup.signup"
                    description="Button label for initiating sign up"
                    defaultMessage="Sign Up"
                  />
                </Button>
              </Grid>
            </form>
          </Grid>
        </Grid>
      </>
    );
  }
}

SignUp.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
  updateUser: PropTypes.func.isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default withCookies(injectIntl(connect(null, mapDispatchToProps)(SignUp)));
