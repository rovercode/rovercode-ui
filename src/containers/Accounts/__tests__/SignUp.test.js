import React from 'react';
import { Redirect } from 'react-router';
import { ListItemText, TextField } from '@material-ui/core';
import { Alert } from '@material-ui/lab';
import { Cookies } from 'react-cookie';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import configureStore from 'redux-mock-store';
import jwtDecode from 'jwt-decode';

import SignUp from '../SignUp';

jest.mock('@/actions/user');

import { updateUser } from '@/actions/user'; // eslint-disable-line import/first, import/order

const mockStore = configureStore();
const store = mockStore();
store.dispatch = jest.fn();

const mock = new MockAdapter(axios);

const cookiesValues = { };
const cookies = new Cookies(cookiesValues);

test('SignUp renders on the page with no errors', () => {
  const cookiesWrapper = shallowWithIntl(<SignUp store={store} />, {
    context: { cookies },
  });

  const wrapper = cookiesWrapper.dive().dive().dive().dive()
    .dive()
    .dive();

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find(Alert).exists()).toBe(false);
});

test('SignUp displays form errors', (done) => {
  mock.reset();
  mock.onPost('/jwt/auth/registration/').reply(400, {
    username: ['A user with that username already exists.'],
    email: ['Enter a valid email address.'],
    password1: [
      'This password is too short. It must contain at least 8 characters.',
      'This password is too common.',
      'This password is entirely numeric.',
    ],
  });
  const cookiesWrapper = shallowWithIntl(<SignUp store={store} />, {
    context: { cookies },
  });

  const wrapper = cookiesWrapper.dive().dive().dive().dive()
    .dive()
    .dive();
  wrapper.instance().signUp({ preventDefault: jest.fn() }).then(() => {
    wrapper.update();

    expect(wrapper.find(Alert).exists()).toBe(true);
    expect(wrapper.find(ListItemText).length).toBe(5);
    expect(wrapper.find(ListItemText).at(0).prop('children')).toBe('- A user with that username already exists.');
    expect(wrapper.find(ListItemText).at(1).prop('children')).toBe('- Enter a valid email address.');
    expect(wrapper.find(ListItemText).at(2).prop('children')).toBe('- This password is too short. It must contain at least 8 characters.');
    expect(wrapper.find(ListItemText).at(3).prop('children')).toBe('- This password is too common.');
    expect(wrapper.find(ListItemText).at(4).prop('children')).toBe('- This password is entirely numeric.');
    expect(wrapper.find(TextField).at(0).prop('error')).toBeTruthy();
    expect(wrapper.find(TextField).at(1).prop('error')).toBeTruthy();
    expect(wrapper.find(TextField).at(2).prop('error')).toBeTruthy();
    expect(wrapper.find(TextField).at(3).prop('error')).toBeFalsy();
    done();
  });
});

test('SignUp redirects to root after success', (done) => {
  const token = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoxLCJ1c2VybmFtZSI6ImFkbWluIiwiZXhwIjoxNTQwMzQzMjIxLCJlbWFpbCI6ImFkbWluQGV4YW1wbGUuY29tIiwib3JpZ19pYXQiOjE1NDAzMzk2MjF9.tumcSSAbKeWXc2QDd7KFR9IGh3PCsyHnCe6JLSszWpc';
  const username = 'admin';
  const email = 'admin@example.com';
  const password1 = 'password123';
  const password2 = 'password123';

  mock.reset();
  mock.onPost('/jwt/auth/registration/', {
    username,
    email,
    password1,
    password2,
  }).reply(200, {
    access_token: token,
  });
  const cookiesWrapper = shallowWithIntl(<SignUp store={store} />, {
    context: { cookies },
  });

  const wrapper = cookiesWrapper.dive().dive().dive().dive()
    .dive()
    .dive();

  wrapper.find(TextField).at(0).simulate('change', {
    target: {
      value: username,
    },
  });
  wrapper.find(TextField).at(1).simulate('change', {
    target: {
      value: email,
    },
  });
  wrapper.find(TextField).at(2).simulate('change', {
    target: {
      value: password1,
    },
  });
  wrapper.find(TextField).at(3).simulate('change', {
    target: {
      value: password2,
    },
  });

  wrapper.instance().signUp({ preventDefault: jest.fn() }).then(() => {
    wrapper.update();

    expect(cookies.get('auth_jwt')).toBe(token);
    expect(wrapper.find(Redirect).exists()).toBe(true);
    expect(wrapper.find(Redirect).prop('to')).toEqual({
      pathname: '/',
    });
    expect(store.dispatch).toHaveBeenCalledWith(
      updateUser({ ...jwtDecode(token), isSocial: false }),
    );

    cookies.remove('auth_jwt');
    done();
  });
});

test('SignUp redirects to next after success', (done) => {
  const token = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoxLCJ1c2VybmFtZSI6ImFkbWluIiwiZXhwIjoxNTQwMzQzMjIxLCJlbWFpbCI6ImFkbWluQGV4YW1wbGUuY29tIiwib3JpZ19pYXQiOjE1NDAzMzk2MjF9.tumcSSAbKeWXc2QDd7KFR9IGh3PCsyHnCe6JLSszWpc';
  const username = 'admin';
  const email = 'admin@example.com';
  const password1 = 'password123';
  const password2 = 'password123';

  mock.reset();
  mock.onPost('/jwt/auth/registration/', {
    username,
    email,
    password1,
    password2,
  }).reply(200, {
    access_token: token,
  });
  const cookiesWrapper = shallowWithIntl(<SignUp store={store} location={{ state: { next: '/courses' } }} />, {
    context: { cookies },
  });

  const wrapper = cookiesWrapper.dive().dive().dive().dive()
    .dive()
    .dive();

  wrapper.find(TextField).at(0).simulate('change', {
    target: {
      value: username,
    },
  });
  wrapper.find(TextField).at(1).simulate('change', {
    target: {
      value: email,
    },
  });
  wrapper.find(TextField).at(2).simulate('change', {
    target: {
      value: password1,
    },
  });
  wrapper.find(TextField).at(3).simulate('change', {
    target: {
      value: password2,
    },
  });

  wrapper.instance().signUp({ preventDefault: jest.fn() }).then(() => {
    wrapper.update();

    expect(cookies.get('auth_jwt')).toBe(token);
    expect(wrapper.find(Redirect).exists()).toBe(true);
    expect(wrapper.find(Redirect).prop('to')).toEqual({
      pathname: '/courses',
    });
    expect(store.dispatch).toHaveBeenCalledWith(
      updateUser({ ...jwtDecode(token), isSocial: false }),
    );

    cookies.remove('auth_jwt');
    done();
  });
});
