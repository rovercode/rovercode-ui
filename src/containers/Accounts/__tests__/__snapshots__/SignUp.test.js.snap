// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp renders on the page with no errors 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    direction="column"
    justify="center"
    spacing={4}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <WithStyles(ForwardRef(Typography))
        variant="h4"
      >
        <FormattedMessage
          defaultMessage="Sign Up"
          id="app.signup.header"
        />
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <WithStyles(ForwardRef(Typography))>
        <FormattedMessage
          defaultMessage="Already have an account? Then please"
          id="app.signup.check_1"
        />
         
        <WithStyles(ForwardRef(Link))
          component={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "Link",
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          to="/accounts/login"
        >
          <FormattedMessage
            defaultMessage="sign in."
            id="app.signup.check_2"
          />
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <form
        onSubmit={[Function]}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(TextField))
            InputProps={
              Object {
                "startAdornment": <WithStyles(ForwardRef(InputAdornment))
                  position="start"
                >
                  <Memo(PersonIcon) />
                </WithStyles(ForwardRef(InputAdornment))>,
              }
            }
            error={false}
            fullWidth={true}
            margin="dense"
            onChange={[Function]}
            placeholder="Username"
            required={true}
            variant="outlined"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(TextField))
            InputProps={
              Object {
                "startAdornment": <WithStyles(ForwardRef(InputAdornment))
                  position="start"
                >
                  <Memo(MailIcon) />
                </WithStyles(ForwardRef(InputAdornment))>,
              }
            }
            error={false}
            fullWidth={true}
            margin="dense"
            onChange={[Function]}
            placeholder="Email"
            required={true}
            variant="outlined"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(TextField))
            InputProps={
              Object {
                "startAdornment": <WithStyles(ForwardRef(InputAdornment))
                  position="start"
                >
                  <Memo(LockIcon) />
                </WithStyles(ForwardRef(InputAdornment))>,
              }
            }
            error={false}
            fullWidth={true}
            margin="dense"
            onChange={[Function]}
            placeholder="Password"
            required={true}
            type="password"
            variant="outlined"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(TextField))
            InputProps={
              Object {
                "startAdornment": <WithStyles(ForwardRef(InputAdornment))
                  position="start"
                >
                  <Memo(LockIcon) />
                </WithStyles(ForwardRef(InputAdornment))>,
              }
            }
            error={false}
            fullWidth={true}
            margin="dense"
            onChange={[Function]}
            placeholder="Password"
            required={true}
            type="password"
            variant="outlined"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            type="submit"
            variant="contained"
          >
            <FormattedMessage
              defaultMessage="Sign Up"
              id="app.signup.signup"
            />
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </form>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</Fragment>
`;
